src/erc721.rs

```diff
@@ -0,0 +1,367 @@
+
+        <OwnedTokensCount<T>>::insert(&owner, new_balance_of);
+        <TokenOwner<T>>::remove(token_id);
+
+        Self::deposit_event(RawEvent::Transfer(Some(owner), None, *token_id));
+
+        Ok(())
+    }
+
+    // Transfer token from one account to other
+    pub(crate) fn _transfer_from(
+        from: &T::AccountId,
+        to: &T::AccountId,
+        token_id: &T::Hash,
+    ) -> DispatchResult {
+        let owner = match Self::owner_of(token_id) {
+            Some(c) => c,
+            None => return Err(Error::<T>::TokenNoOwner.into()),
+        };
+
+        ensure!(owner == *from, "'from' account does not own this token");
+
+        let balance_of_from = Self::balance_of(from);
+        let balance_of_to = Self::balance_of(to);
+
+        let new_balance_of_from = match balance_of_from.checked_sub(1) {
+            Some(c) => c,
+            None => return Err(Error::<T>::Underflow.into()),
+        };
+
+        let new_balance_of_to = match balance_of_to.checked_add(1) {
+            Some(c) => c,
+            None => return Err(Error::<T>::Overflow.into()),
+        };
+
+        Self::_remove_token_from_owner_enumeration(from, token_id)?;
+        Self::_add_token_to_owner_enumeration(to, token_id)?;
+
+        Self::_clear_approval(token_id)?;
+        <OwnedTokensCount<T>>::insert(from, new_balance_of_from);
+        <OwnedTokensCount<T>>::insert(to, new_balance_of_to);
+        <TokenOwner<T>>::insert(token_id, to);
+
+        Self::deposit_event(RawEvent::Transfer(
+            Some(from.clone()),
+            Some(to.clone()),
+            *token_id,
+        ));
+
+        Ok(())
+    }
+
+    // Clear one token's delegate
+    fn _clear_approval(token_id: &T::Hash) -> DispatchResult {
+        <TokenApprovals<T>>::remove(token_id);
+        Ok(())
+    }
+
+    // Add token index to its owner's enumeration
+    fn _add_token_to_owner_enumeration(to: &T::AccountId, token_id: &T::Hash) -> DispatchResult {
+        let new_token_index = Self::balance_of(to);
+        <OwnedTokensIndex<T>>::insert(token_id, new_token_index);
+        <OwnedTokens<T>>::insert((to, new_token_index), token_id);
+
+        Ok(())
+    }
+
+    // Add token index to all token's enumeration
+    fn _add_token_to_all_tokens_enumeration(token_id: &T::Hash) -> DispatchResult {
+        let total_supply = Self::total_supply();
+
+        let new_total_supply = match total_supply.checked_add(1) {
+            Some(c) => c,
+            None => return Err(Error::<T>::Overflow.into()),
+        };
+
+        let new_token_index = total_supply;
+
+        <AllTokensIndex<T>>::insert(token_id, new_token_index);
+        <AllTokens<T>>::insert(new_token_index, token_id);
+        TotalSupply::put(new_total_supply);
+
+        Ok(())
+    }
+
+    // Remove token info from owner
+    fn _remove_token_from_owner_enumeration(
+        from: &T::AccountId,
+        token_id: &T::Hash,
+    ) -> DispatchResult {
```

Comment:
The index handling here seems fairly complicated for the ability to iterate over tokens. 

In Substrate, you don't need to use indexes for collections, since you can just iterate over the storage by prefix. It might be better to use a double map and insert/remove tokens by their unique ID instead of doing an indirection with indexes.

Reply:
I will try double map to see if bi-direction query can be efficiently realized.